libraries/botbuilder-dialogs-adaptive-testing/src/testRunner.ts

```diff
@@ -29,5 +30,8 @@ export class TestRunner {
         script.description = script.description || testName;
         this.testAdapter.activeQueue = [];
         await script.execute(this.resourceExplorer, testName, this.testAdapter);
+        // We should clear the event listeners between each test. Avoids test pollution and MaxListenersExceeded warning.
+        // @ts-ignore: Property is private, but we need to reset it, only for testing.
```

Comment:
### Alternative: INPUT NEEDED

Instead of doing this a sort of "hacky" way, we could add a `public` method to `ResourceExplorer` that allows users to reset the event emitter. Something like:

```js
public resetEventEmitter() {
	this._eventEmitter = new EventEmitter();
}
```

Let me know if you'd prefer that route and I can make the adjustment in both `ResourceExplorer` and the tests.

Reply:
@luhan2017, @chrimc62, @boydc2014 could someone comment on this?

Is this an issue that customers would normally run into? Either through the testing package or through using `ResourceExplorer` in their business logic?

If there is merit in developers having access to some of the basic methods on EventEmitter, then we could use a facade to limit direct dependencies.

If it just happens in our tests, then we can develop some internal tools/workarounds to cleanup the TestRunner's underlying ResourceExplorer.

___

@mdrichardson would [EventEmitter.removeAllListeners()](https://nodejs.org/api/events.html#events_emitter_removealllisteners_eventname) suffice instead of creating a new EventEmitter?

> ## `emitter.removeAllListeners([eventName])#`
> Added in: v0.1.26
eventName `<string> | <symbol>`
Returns: `<EventEmitter>`
Removes all listeners, or those of the specified eventName.
>
> It is bad practice to remove listeners added elsewhere in the code, particularly when the EventEmitter instance was created by some other component or module (e.g. sockets or file streams).
>
> Returns a reference to the EventEmitter, so that calls can be chained.

Reply:
@stevengum I considered using `removeAllListeners`, but shied away from it because of the "bad practice" warning. It also seemed like, for avoiding test pollution, it might be best to just use an entirely new instance. `removeAllListeners` *does* work, though.

Reply:
I would note that the ResourceExplorer is the one creating the EventEmitter, so that places a greater emphasis on making that method available from ResourceExplorer and not stamping on top of it in another class. 😄 

Also, I don't believe there is any semantic difference in removing all listeners registered on a private member EventEmitter versus removing replacing it with a new one.

I think once we hear from the architects what the expected/supported interactions with the ResourceExplorer's EventEmitter are, we can decide on an implementation.

Reply:
@luhan2017 @boydc2014 can you chime in on this PR?

Reply:
An alternative approach would be to construct the resource explorer and test adapter within the scope of the `runTestScript` method. This benefits from ensuring that tests are executed with potentially side-effect-free instances (I.e., any side effects of running test script A do not affect test script B).

The downside of this approach is slowing the test suite down. If constructing a resource explorer instance is expensive, then perhaps a reset method on the resource explorer itself makes sense.